IdentityCore/src/cache/MSIDMacKeychainTokenCache.m

```diff
@@ -0,0 +1,660 @@
+                                                        serializer:(__unused id<MSIDAppMetadataItemSerializer>)serializer
+                                                           context:(__unused id<MSIDRequestContext>)context
+                                                             error:(__unused NSError **)error
+{
+    [self createUnimplementedError:error context:context];
+    return nil;
+}
+
+// Remove items with the given Metadata key from the macOS keychain cache.
+- (BOOL)removeItemsWithMetadataKey:(__unused MSIDCacheKey *)key
+                           context:(__unused id<MSIDRequestContext>)context
+                             error:(NSError **)error
+{
+    [self createUnimplementedError:error context:context];
+    return FALSE;
+}
+
+#pragma mark - Wipe Info
+
+// Saves information about the app which most-recently removed a token.
+- (BOOL)saveWipeInfoWithContext:(__unused id<MSIDRequestContext>)context
+                          error:(NSError **)error
+{
+    [self createUnimplementedError:error context:context];
+    return FALSE;
+}
+
+// Read information about the app which most-recently removed a token.
+- (NSDictionary *)wipeInfo:(__unused id<MSIDRequestContext>)context
+                     error:(NSError **)error
+{
+    [self createUnimplementedError:error context:context];
+    return nil;
+}
+
+#pragma mark - clear
+
+// A test-only method that deletes all items from the cache for the given context.
+- (BOOL)clearWithContext:(id<MSIDRequestContext>)context
+                   error:(NSError **)error
+{
+    MSID_TRACE;
+    MSID_LOG_WARN(context, @"Clearing the whole context. This should only be executed in tests");
+
+    // For now, this just deletes all accounts (an empty key matches all account items):
+    MSIDMacKeychainAccountCacheKey* key = [MSIDMacKeychainAccountCacheKey new];
+    return [self removeItemsWithAccountKey:key context:context error:error];
+}
+
+#pragma mark - Utilities
+
+// Get the basic/default keychain query dictionary for account items.
+// To support searching, the key properties may be partially or fully empty.
+- (NSMutableDictionary *)accountQueryForKey:(MSIDCacheKey *)key
+{
+    MSID_TRACE;
+    NSMutableDictionary *query = [self.defaultAccountQuery mutableCopy];
+
+    if (key.account.length > 0)
+    {
+        // Account attribute: keychainGroup>-<homeAccountId>-<environment>
+        // The access group is added to the account attribute so it's part of the keychain item's primary key.
+        query[(id)kSecAttrAccount] = [NSString stringWithFormat:@"%@-%@", self.keychainGroup, key.account];
+    }
+    if (key.service.length > 0)
+    {
+        // Service attribute: <realm>
+        query[(id)kSecAttrService] = key.service;
+    }
+
+    return query;
+}
+
+// Get the basic/default keychain update dictionary for account items.
+// These are not _primary_ keys, but if they're present in the key object we
+// want to set them when adding/updating.
```

Comment:
We also want to be able to query by them, not only updating.

Reply:
updated